src/lib/system_logger.ts

```diff
@@ -0,0 +1,115 @@
+const systemLogTag = '__nfSystemLog'
+
+const serializeError = (error: Error): Record<string, unknown> => {
+  const cause = error?.cause instanceof Error ? serializeError(error.cause) : error.cause
+
+  return {
+    error: error.message,
+    error_cause: cause,
+    error_stack: error.stack,
+  }
+}
+
+// eslint pretends there's a different enum at the same place - it's wrong!
+// eslint-disable-next-line no-shadow
+export enum LogLevel {
+  Debug = 1,
+  Info,
+  Warn,
+  Error,
+}
+
+class SystemLogger {
+  // eslint-disable-next-line no-useless-constructor
+  constructor(
+    private readonly fields: Record<string, unknown> = {},
+    private readonly logLevel = LogLevel.Info,
+    private readonly samplingRate = 1,
+  ) {}
+
+  private doLog(logger: typeof console.log, message: string) {
+    if (this.samplingRate < 1 && Math.random() > this.samplingRate) {
+      return
+    }
+
+    logger(systemLogTag, JSON.stringify({ msg: message, fields: this.fields }))
+  }
+
+  /**
+   * Alias for .info
+   */
+  log(message: string) {
+    this.info(message)
+  }
+
+  info(message: string) {
+    if (this.logLevel > LogLevel.Info) {
+      return
+    }
+
+    this.doLog(console.info, message)
+  }
+
+  debug(message: string) {
+    if (this.logLevel > LogLevel.Debug) {
+      return
+    }
+
+    this.doLog(console.debug, message)
+  }
+
+  warn(message: string) {
+    if (this.logLevel > LogLevel.Warn) {
+      return
+    }
+
+    this.doLog(console.warn, message)
+  }
+
+  error(message: string) {
+    if (this.logLevel > LogLevel.Error) {
+      return
+    }
+
+    this.doLog(console.error, message)
+  }
+
+  withLogLevel(level: LogLevel) {
+    return new SystemLogger(this.fields, level, this.samplingRate)
+  }
+
+  withSamplingRate(rate: number) {
+    return new SystemLogger(this.fields, this.logLevel, rate)
+  }
+
+  withFields(fields: Record<string, unknown>) {
+    return new SystemLogger(
+      {
+        ...this.fields,
+        ...fields,
+      },
+      this.logLevel,
+      this.samplingRate,
+    )
+  }
+
+  withError(error: unknown) {
+    const fields = error instanceof Error ? serializeError(error) : { error }
+
+    return this.withFields(fields)
+  }
+
+  withRequest(req: Request) {
```

Comment:
Edge-Utils only has a `withRequestID` function, not a `withRequest` one. I've opened a PR over there to also add `withRequest`, so both our loggers adhere to the same shape: https://github.com/netlify/edge-utils/pull/36